quinn-proto/src/connection/mtud.rs

```diff
@@ -336,132 +335,174 @@ impl SearchState {
+    /// no larger packets have been received since the most recent loss burst.
+    acked_mtu: u16,
     /// The UDP payload size guaranteed to be supported by the network
     min_mtu: u16,
 }
 
 impl BlackHoleDetector {
     fn new(min_mtu: u16) -> Self {
         Self {
-            suspicious_loss_bursts: 0,
-            largest_non_probe_lost: None,
-            loss_burst_has_non_suspicious_packets: false,
-            largest_acked_mtu_sized_packet: None,
+            suspicious_loss_bursts: Vec::with_capacity(BLACK_HOLE_THRESHOLD + 1),
+            current_loss_burst: None,
+            largest_post_loss_packet: 0,
+            acked_mtu: min_mtu,
             min_mtu,
         }
     }
 
-    fn on_probe_acked(&mut self) {
-        // We know for sure the path supports the current MTU
-        self.suspicious_loss_bursts = 0;
+    fn on_probe_acked(&mut self, pn: u64, len: u16) {
+        debug_assert!(
+            pn >= self.largest_post_loss_packet,
+            "ACKs are delivered in order"
+        );
+        // MTU probes are always larger than the previous MTU, so no previous loss bursts are
+        // suspicious.
+        self.suspicious_loss_bursts.clear();
+        self.acked_mtu = len;
+        self.largest_post_loss_packet = pn;
     }
 
-    fn on_non_probe_acked(&mut self, current_mtu: u16, pn: u64, len: u16) {
-        // Reset the black hole counter if a packet the size of the current MTU or larger
-        // has been acknowledged
-        if len >= current_mtu && self.largest_acked_mtu_sized_packet.map_or(true, |x| pn > x) {
-            self.suspicious_loss_bursts = 0;
-            self.largest_acked_mtu_sized_packet = Some(pn);
+    fn on_non_probe_acked(&mut self, pn: u64, len: u16) {
+        debug_assert!(
+            pn >= self.largest_post_loss_packet,
+            "ACKs are delivered in order"
+        );
+        if len <= self.acked_mtu {
+            // We've already seen a larger packet since the most recent suspicious loss burst;
+            // nothing to do.
+            return;
         }
+        self.acked_mtu = len;
+        self.largest_post_loss_packet = pn;
+        // Loss bursts packets smaller than this are retroactively deemed non-suspicious.
+        self.suspicious_loss_bursts
+            .retain(|burst| burst.smallest_packet_size > len);
     }
 
     fn on_non_probe_lost(&mut self, pn: u64, len: u16) {
         // A loss burst is a group of consecutive packets that are declared lost, so a distance
         // greater than 1 indicates a new burst
-        let new_loss_burst = self
-            .largest_non_probe_lost
-            .map_or(true, |prev| pn - prev != 1);
+        let end_last_burst = self
+            .current_loss_burst
+            .as_ref()
+            .map_or(false, |current| pn - current.latest_non_probe != 1);
 
-        if new_loss_burst {
+        if end_last_burst {
```

Comment:
I wonder if the interface could be simplified here? In https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/connection/mod.rs#L1673 we have a loop that calls `on_non_probe_lost()` for each packet that was deemed lost, which I think can only add suspicious loss bursts. After the loop we call `black_hole_detected()`, but it feels like we could conceivably yield a value indicating a black hole from `on_non_probe_lost()` directly, simplifying the internal structure by inlining the `black_hole_detected()` logic into `finish_loss_burst()` (and maybe `finish_loss_burst()` into `on_non_probe_lost()`).

Reply:
Interesting idea. I had a look and I think this may be less straightforward than it appears. The current shape makes the termination of that loop visible to black hole detection logic. This allows us to avoid counting losses towards a future black hole detection after we've already detected a black hole due to earlier losses implied by the same ACK, reducing some false positive risk. Clearing the suspicious loss burst list only after processing all losses is what enables this, and wouldn't be feasible if we merged all the logic into `on_non_probe_lost`.

Insofar as that there's room for improvement here, perhaps leave that to future work?